packages/openapi-v3/src/build-responses-from-metadata.ts

```diff
@@ -0,0 +1,150 @@
+          } else {
+            return m;
+          }
+        }),
+      },
+    };
+  } else {
+    const modelOrSpec = Array.isArray(modelOrSpecs)
+      ? modelOrSpecs[0]
+      : modelOrSpecs;
+    if (isModel(modelOrSpec)) {
+      specContents[contentType] = {
+        schema: {'x-ts-type': modelOrSpec},
+      };
+    } else {
+      specContents[contentType] = {
+        schema: modelOrSpec,
+      };
+    }
+  }
+  return specContents;
+}
+
+/**
+ * Reducer which builds the content sections of the operation responses
+ */
+function reduceSpecResponses(
+  specResponses: ResponseObject,
+  [responseCode, c]: [number, {description: string; content: ContentMap}],
+): ResponseObject {
+  const responseContent = c.content;
+  // check if there is an existing block, from something like an inhered @op spec
+  if (Object.prototype.hasOwnProperty.call(specResponses, responseCode)) {
+    // we might need to merge
+    const content = Array.from(responseContent).reduce(
+      reduceSpecContent,
+      specResponses[responseCode].content as ContentObject,
+    );
+
+    specResponses[responseCode] = {
+      description: c.description,
+      content,
+    };
+  } else {
+    const content = Array.from(responseContent).reduce(
+      reduceSpecContent,
+      {} as ContentObject,
+    );
+
+    specResponses[responseCode] = {
+      description: c.description,
+      content,
+    };
+  }
+  return specResponses;
+}
+
+/**
+ * This function takes an array of flat-ish data:
+ * ```
+ * [
+ *  { responseCode, contentType, description, modelOrSpec },
+ *  { responseCode, contentType, description, modelOrSpec },
+ * ]
+ * ```
+ * and turns it into a multi-map structure that more closely aligns with
+ * the final json
+ * ```
+ * Map{ [code, Map{[contentType, modelOrSpec], [contentType, modelOrSpec]}]}
+ * ```
+ */
+function buildMapsFromMetadata(
+  metadata: ResponseDecoratorMetadata,
+): ResponseMap {
+  const builder: ResponseMap = new Map();
+  metadata.forEach(r => {
+    if (builder.has(r.responseCode)) {
+      const responseRef = builder.get(r.responseCode);
+      const codeRef = responseRef?.content;
+
+      if (codeRef?.has(r.contentType)) {
+        // eslint-disable-next-line no-unused-expressions
+        codeRef.get(r.contentType)?.push(r.responseModelOrSpec);
+      } else {
+        // eslint-disable-next-line no-unused-expressions
+        codeRef?.set(r.contentType, [r.responseModelOrSpec]);
+      }
+    } else {
+      const codeRef = new Map();
+      codeRef.set(r.contentType, [r.responseModelOrSpec]);
+      builder.set(r.responseCode, {
+        description: r.description,
+        content: codeRef,
+      });
+    }
+  });
+  return builder;
+}
+export function buildResponsesFromMetadata(
+  metadata: ResponseDecoratorMetadata,
+  existingOperation?: OperationObject,
+): OperationObject {
+  const builder = buildMapsFromMetadata(metadata);
+  const base = existingOperation
+    ? _.cloneDeep(existingOperation.responses)
```

Comment:
We have `cloneDeep` from `DecoratorFactory`.

Reply:
:+1: